### PR TITLE
Fix toleration comparison & merging logic

### PR DESCRIPTION
**Pull request by tallclair**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

The list of tolerations on a pod is effectively represents a union of the taints tolerated by each, so it is entirely reasonable to have 2 tolerations like:
```yaml
- key: foo
  operator: Equal
  value: bar
- key: foo
  operator: Equal
  value: baz
```

However, the old logic would consider those 2 tolerations a conflict, and overwrite one with the other in a merge.

This PR fixes the toleration helper methods to represent a logical merge and comparison. It does that by recognizing that some tolerations are supersets of others, due to wildcard matches. For example, the above tolerations would be redundant if there was a 3rd toleration: `{"key": "foo", "operator": "Exists"}`.

These fixes are needed to better support [RuntimeClass scheduling](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/runtime-class-scheduling.md), but also affect the alpha PodTolerationRestriction admission controller.

**Special notes for your reviewer**:

There are 2 design decisions I want to highlight, and am open to discussion on:

1. The decision to "smart merge" - Since tolerations represent a union, we could perform the merge by simply appending one list to the other (and maybe eliminating exact duplicates). We still need the `isSuperset` logic for checking the whitelist though, and I feel that simplifying with the smart merge offers a cleaner user experience.
2. `TolerationSeconds` - The [eviction logic](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodelifecycle/scheduler/taint_manager.go#L132) actually takes the minimum toleration time. So technically a lower toleration time would be the superset of a higher time. However, I don't think that was the intention (that logic probably wasn't considering duplicate tolerations for the same taint, but rather across all taints), so I've stuck with using the maximum toleration time for the superset. I think that is more inline with the intention of both the PodTolerationRestriction and RuntimeClass admission controllers (probably irrelevant for the latter).

**Does this PR introduce a user-facing change?**:
```release-note
Fix an issue with toleration merging & whitelist checking in the PodTolerationRestriction admission controller.
```

/sig scheduling
/sig node
/priority important-soon
/milestone v1.16

I wasn't sure who to assign this to, so went with the original author and reviewers:
/assign @aveshagarwal 
/cc @timothysc @smarterclayton @derekwaynecarr @davidopp @liggitt

### CODE REVIEW

**Comment by tallclair** [comment]
Forgot to update PodTolerationRestriction. Fixed.

**Comment by tallclair** [comment]
Rebased.

**Comment by mattjmcnaughton** [comment]
/remove-sig node

Unless I'm missing something, not sure this touches sig-node code. Please feel free to add the sig back if I'm incorrect :)

File: pkg/util/tolerations/tolerations.go [hunk 1]
```diff
@@ -17,114 +17,91 @@ limitations under the License.
...
-	// preserve order of first when merging
-	for _, v := range first {
-		k := convertTolerationToKey(v)
-		// if first contains key conflicts, the last one takes precedence
-		if _, ok := secondMap[k]; !ok && firstMap[k] == v {
-			mergedTolerations = append(mergedTolerations, v)
+		if i+1 < len(all) {
+			for _, t2 := range all[i+1:] {
+				// If the tolerations are equal, prefer the first.
+				if isSuperset(t2, t) && !apiequality.Semantic.DeepEqual(&t, &t2) {
```
Comment (tedyu) [inline pkg/util/tolerations/tolerations.go]: apiequality.Semantic.DeepEqual is called by isSuperset.

I think the order of the conditions should be :
```
				if !apiequality.Semantic.DeepEqual(&t, &t2) && isSuperset(t2, t) {
```
This way, if apiequality.Semantic.DeepEqual(&t, &t2) == true, isSuperset won't be called.

**Review by draveness — APPROVED** [review]
/lgtm

**Comment by draveness** [comment]
/retest

Regards,
Draven

**Pull request merged**